### PR TITLE
WIP: fix(typedefs): fix `source` being resolved as `context`

### PR DESCRIPTION
**Pull request by mernxl**
Adoptive commit

When using Mongoose plugin, i often get the source interface in the context arg of resolvers so i decided to look into the issue and brought this up. Please review and ask any question.

@nodkz @toverux you might want to check this changes...
This is just a test commit for review to see if i am getting the idea right.

```typescript
import { SchemaComposer } from '../index';
import { TypeComposer } from '../TypeComposer';

interface Person {
    uid: string;
    name: string;
    nickName: string;
    age: number;
}

interface Context {
    req: any;
    res: any;
    uid: string;
}

// works for cases where u create your instance of
const schemaComposer = new SchemaComposer<Context>();

// Person from this schemaComposer gets the context defined for that schemaComposer
const Person = schemaComposer.getOrCreateTC<Person>('Person');

// create from default schemaComposer uses the context type passed in
// or uses `any` as type
const PersonTC = TypeComposer.create<Person, Context>({
    name: 'Person',
    fields: {
        name: 'String',
        nickName: 'String',
        age: 'Int'
    }
});

PersonTC.addResolver({
    resolve: ({ source, context }) => {
        // check availability
        if (source && context) {
            source.age = 2;
            source.name = 'XYZ';
            if (context.uid !== source.uid) {
                // do something if not the current user ...
            }
        }
    }
});

// Changes a driven by the quest for better typing and a better Intelisense experience
// hence boosting productivity
// It passes linting
```

The above passed the `tscheck`. I also noticed that this TContext type implementations is also used at the FlowType end. Well If we are able to come to a consensus concerning this new implementation, then we can go ahead to convert the Flow types.

### CODE REVIEW

**Comment by nodkz** [comment]
We can use `TContext` for `TypeComposer` but not `TSource`.
- `TContext` is global and does not change, so it's good to provide it as generic
- `TSource` has a parent object data, according to `graphql-js`
  - for fields, it will be the current model
  - BUT, for `Resolvers` it will be **not the current model**. Resolver is a `FieldConfig` for some other type. And `TSource` for Resolver depends on where it will be used.

Simple demo:
```js
PersonTC.addResolver({
    findMany: ({ source: TSourceDependsOnWhereItUsed, context: TContext }) => {}
});

// here TSourceDependsOnWhereItUsed will be TQuerySource;
schemaComposer.Query.addFields({
  personMany: PersonTC.getResolver('findMany'), 
});

// here TSourceDependsOnWhereItUsed will be TBookingSource;
BookingTC.addRelation('persons', {
  resolver: () => PersonTC.getResolver('findMany'),
  prepareArgs: {
    _ids: source => source.personIds,
  },
  projection: { personIds: true },
});
```

But for fields it will be OK:
```js
PersonTC.addFields({
  calculatedRating: (source, args, context, info) => { 
    // `source` is definitely is TSourcePerson
  }
});
```

But for this case you may provide type for source explicitly, like so:
```js
PersonTC.addFields({
  calculatedRating: (source: TSourcePerson, args, context, info) => {}
});
```

Mess with `fields` and `resolvers` stopped me in past to define somehow `TSource` in `TypeComposer`.

**Comment by mernxl** [comment]
I understand `TContext` is global that's why you see me passing through a custom Schema composer like so;
```js
const gqc = new SchemaComposer<Context>();
```
What happened there's that `Context` will be pass to all resolvers and TypeCompose that are created with that `gqc`.
I was actually reasoning each TypeCompose is actually a `Type` which can have an interface, that interface being passed in governs the manipulation of that type ensuring you are type safe. 
But for resolvers not linked to that `Type`, you may define it's source and pass in through `TSource`. I propose this cuz of type safety...

File: src/__tests__/typescript-typedef.spec.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+const PersonTC = TypeComposer.create<Person, Context>({
+    name: 'Person',
+    fields: {
+        name: 'String',
+        nickName: 'String',
+        age: 'Int'
+    }
+});
+
+PersonTC.addResolver({
```
Comment (nodkz) [inline src/__tests__/typescript-typedef.spec.ts]: Inside `addResolver()` method the `source` cannot be equal to `Person`.
How I said before, `resolvers` are fieldConfigs for the external types.

Need to change this test case for `setFields({ ... })`.
`resolvers` should be unaffected by provided `TSource` to TypeComposer (it still should be any for better times ;)).

**Comment by nodkz** [comment]
1) Need fix in test: replace `addResolver` by `setFields`.
2) And let's look how `const HeyRolver = PersonTC.getResolver<TResolverSource>('heyResolver');` works in tests. So needs to add a test case. We should call two times the same `getResolver` with different interfaces. And check how ts works inside `Resolver.resolve` method for polymorphic interfaces.

**Comment by mernxl** [comment]
I don't quite understand point 1. But you look at the commit updates... It may have solved the point.

**Comment by mernxl** [comment]
You will see comments in the test file. Few reasons for my thinking.

**Comment by nodkz** [comment]
@mernxl I added a commit with changes in test file. Please take care of fixing libdefs for removing this errors.

Hope it brings some clearance.

**Comment by mernxl** [comment]
Hello @nodkz, i added a `TArgs` to resolvers so it also helps with type checking of args.
Awaiting review...

**Comment by mernxl** [comment]
Thanks @nodkz, I was thinking I could make same kinda changes to flow types.

**Comment by mernxl** [comment]
Will work the trials and put up for cross checks

**Comment by nodkz** [comment]
Will be nice to get such flow checks. But it can be a quite complex task, cause it will require full compatibility with graphql-js package.

So will be nice to do it in separate PR. It will not block us to merge current chages if work with flow will be delayed due complexity. So if you met problems with, feel to commit a partial work. I'll help you with pleasure.

**Comment by mernxl** [comment]
hey @nodkz, I have tried working out a similar implementation to `flow types` but had some shortcomings, maybe we just discourse them here before i go ahead to create(or not) a PR for that task.

1. Unable to map type names to their types for use in `TArgs`
```typescript
// typescript supports this
export type ResolveParams<TSource, TContext, TArgs = any> = {
  source: TSource;
  args: { [argName in keyof TArgs]: TArgs[argName] };   // <-- Here i can map keys to types
  context: TContext;
  info: graphql.GraphQLResolveInfo;
  projection: Partial<ProjectionType<TSource>>;
  [opt: string]: any;
};
```
But with flow, I am able to support just the following, which kind of doesn't really go well.
```js
export type ResolveParams<TSource, TContext> = {
  source: TSource,
  args: { [argName: string]: any },    // <-- Could not map as of typescript
  context: TContext,
  info: GraphQLResolveInfo,
  projection: $Shape<ProjectionType>,
  [opt: string]: any,
};
```
This point very much cancels out the addition of `TArgs`.

2. For some reason i don't understand, `generic defaults` are throwing errors. But from [flow docs](https://flow.org/en/docs/types/generics), I seem to be on track. It seem to be an `Eslint` issue though...
```js
export class TypeComposer<TSource = any, TContext = any> {   // somehow, this works
...
203  static createTemp<TSrc = any, TCtx = any>(    // this eslint errors with message below
204      opts: TypeAsString | ComposeObjectTypeConfig<TSrc, TCtx> | GraphQLObjectType
205    ): TypeComposer<TSrc, TCtx>
206  if (!this.schemaComposer) {
...
}
```
```
ESLint: Parsing error: Unexpected token 201 | } 202 | > 203 | static createTemp<TSrc = any, TCtx = any>( | ^ 204 | opts: TypeAsString | ComposeObjectTypeConfig<TSrc, TCtx> | GraphQLObjectType 205 | ): TypeComposer<TSrc, TCtx> { 206 | if (!this.schemaComposer) {
```
This issue will really go along way to distort the implementation. So i will be looking forward for suggestions and ways forward. Haven't opened an issue to `eslint-plugin-flowtype`.
Without the above `generic defaults`, I will need to use `overloads` which i am not sure will go quite well as the codebase is in `flow`.

Looking forward for a solution. Meantime i will be working on `TypeScript types` for `...compose-mongoose`.

**Comment by mernxl** [comment]
I will go ahead and open an issue at `eslint-plugin-flowtype`...

**Pull request merged**